nomos-da/network/core/src/protocols/dispersal/validator/behaviour.rs

```diff
@@ -145,3 +145,431 @@ impl<M: MembershipHandler<Id = PeerId, NetworkId = SubnetworkId> + 'static> Netw
+                                 msg_0_counter += 1;
+                             } else {
+                                 msg_1_counter += 1;
+                             }
+                         }
+                    }
+                }
+
+                _ = time::sleep(Duration::from_secs(2)) => {
+                    warn!("Validator timeout reached");
+                    break;
+                }
+            }
+        }
+        (msg_0_counter, msg_1_counter)
+    }
+
+    async fn send_dispersal_messages(
+        disperse_blob_sender: UnboundedSender<(u32, DaBlob)>,
+        subnet_id: u32,
+        messages_to_send: usize,
+    ) {
+        for i in 0..messages_to_send {
+            debug!("Sending blob {i} to subnet {subnet_id} ...");
+            disperse_blob_sender
+                .send((
+                    subnet_id,
+                    DaBlob {
+                        column_idx: 0,
+                        column: Column(vec![]),
+                        column_commitment: Default::default(),
+                        aggregated_column_commitment: Default::default(),
+                        aggregated_column_proof: Default::default(),
+                        rows_commitments: vec![],
+                        rows_proofs: vec![],
+                    },
+                ))
+                .unwrap();
+        }
+    }
+
+    #[tokio::test]
+    async fn test_validation_behaviour() {
+        let _ = tracing_subscriber::fmt()
+            .with_env_filter(EnvFilter::from_default_env())
+            .compact()
+            .with_writer(TestWriter::default())
+            .try_init();
+
+        const ALL_INSTANCES: usize = 20;
+        const MESSAGES_TO_SEND: usize = 10;
+        const GROUPS: usize = 4;
+
+        let executor_0_config = prepare_swarm_config(ALL_INSTANCES / GROUPS, 0);
+        let validator_0_config = prepare_swarm_config(ALL_INSTANCES / GROUPS, 1);
+
+        let executor_1_config = prepare_swarm_config(ALL_INSTANCES / GROUPS, 2);
+        let validator_1_config = prepare_swarm_config(ALL_INSTANCES / GROUPS, 3);
+
+        let subnet_0_ids = executor_0_config
+            .iter()
+            .chain(validator_0_config.iter())
+            .map(|(_, peer_id, _)| *peer_id)
+            .collect::<Vec<_>>();
+
+        let subnet_1_ids = executor_1_config
+            .iter()
+            .chain(validator_1_config.iter())
+            .map(|(_, peer_id, _)| *peer_id)
+            .collect::<Vec<_>>();
+
+        let to_p2p_address = |(_, peer_id, addr): &(_, PeerId, Multiaddr)| {
+            (*peer_id, addr.clone().with_p2p(*peer_id).unwrap())
+        };
+
+        let validator_addressbook = AddressBook::from_iter(
+            validator_0_config
+                .iter()
+                .map(to_p2p_address)
+                .chain(validator_1_config.iter().map(to_p2p_address)),
+        );
+
+        let subnet_0_membership = create_membership(ALL_INSTANCES / 2, 0, &subnet_0_ids);
```

Comment:
This seems a bit strange, ideally we would have one membership with all peers assigned to different subnetworks. Executors would use membership to connect to any peer that is in that subnetwork, including another executor.

Maybe the initialization of different kinds of swarms could be based on one membership, this would simplify the setup and be closer to what we expect in real network?

Reply:
Fixed at https://github.com/logos-co/nomos-node/pull/819/commits/8397a1f043ce64968363950c087cb593d12e1f31

Reply:
About a half of executors won't receive dispersal success message when address book contains all swarms. And the same amount never reaches validator for dispersal. 

Visible from test run with `RUST_LOG=debug`:

```
2024-10-31T02:37:32.180833Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.181661Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.181748Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.181781Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.181829Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.181859Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.271488Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.271625Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.283855Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.283928Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator timeout reached
2024-10-31T02:37:32.284586Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 20 messages from subnet 0
2024-10-31T02:37:32.284597Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284608Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 0
2024-10-31T02:37:32.284614Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 10 messages from subnet 1

2024-10-31T02:37:32.284621Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 0
2024-10-31T02:37:32.284627Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284633Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 0
2024-10-31T02:37:32.284638Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284645Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 10 messages from subnet 0
2024-10-31T02:37:32.284651Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284657Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 0
2024-10-31T02:37:32.284663Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284669Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 0
2024-10-31T02:37:32.284675Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284681Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 0
2024-10-31T02:37:32.284687Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284693Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 0
2024-10-31T02:37:32.284699Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284706Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 0
2024-10-31T02:37:32.284711Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Validator received 0 messages from subnet 1

2024-10-31T02:37:32.284727Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 10 messages dispersal success
2024-10-31T02:37:32.284738Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor terminated
2024-10-31T02:37:32.284794Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor terminated
2024-10-31T02:37:32.284839Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor terminated
2024-10-31T02:37:32.284891Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor terminated
2024-10-31T02:37:32.284931Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor terminated
2024-10-31T02:37:32.284970Z  WARN nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor terminated
2024-10-31T02:37:32.285421Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 0 messages dispersal success
2024-10-31T02:37:32.285431Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 0 messages dispersal success
2024-10-31T02:37:32.285438Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 10 messages dispersal success
2024-10-31T02:37:32.285445Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 0 messages dispersal success
2024-10-31T02:37:32.285452Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 0 messages dispersal success
2024-10-31T02:37:32.285459Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 10 messages dispersal success
2024-10-31T02:37:32.285466Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 0 messages dispersal success
2024-10-31T02:37:32.285472Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 10 messages dispersal success
2024-10-31T02:37:32.285478Z DEBUG nomos_da_network_core::protocols::dispersal::validator::behaviour::tests: Executor task received: 0 messages dispersal success
```

https://github.com/logos-co/nomos-node/pull/819/commits/1121ffabdafe1058c62acfb91529ad18601b742c

Reply:
Reverted to validators only, as having executors within the address book is not supported yet. https://github.com/logos-co/nomos-node/pull/819/commits/d4a20fbe3075955bedf92757602df2559543b160

Reply:
Executors only send stuff. In this case for dispersal, they will not receive anything, never. If one executor behaviour connects to other executor behaviour is a bug.

Reply:
> Executors only send stuff. In this case for dispersal, they will not receive anything, never. If one executor behaviour connects to other executor behaviour is a bug.

Opened https://github.com/logos-co/nomos-node/issues/900